source/grid/grid_tools.cc

```diff
@@ -6452,6 +6458,456 @@ namespace GridTools
+      indices_of_rank[my_rank] = std::vector<unsigned int>();
+
+      // rank to used indices on the rank known on recv side
+      std::map<unsigned int, std::vector<unsigned int>> send_indices_of_rank;
+      for (const auto rank : recv_ranks)
+        if (rank != my_rank)
+          send_indices_of_rank[rank] = std::vector<unsigned int>();
+
+      // fill the maps
+      for (const auto &point_recv_component : point_recv_components)
+        {
+          const auto rank = std::get<0>(point_recv_component);
+          const auto idx  = std::get<1>(point_recv_component);
+
+          if (rank == my_rank)
+            indices_of_rank[rank].emplace_back(idx);
+          else
+            send_indices_of_rank[rank].emplace_back(idx);
+        }
+
+      // send indices to the ranks we normally receive from
+      for (const auto rank : recv_ranks)
+        {
+          if (rank == my_rank)
+            continue;
+
+          auto buffer = Utilities::pack(send_indices_of_rank[rank], false);
+
+          requests.push_back(MPI_Request());
+
+          const int ierr = MPI_Isend(buffer.data(),
+                                     buffer.size(),
+                                     MPI_CHAR,
+                                     rank,
+                                     mpi_tag,
+                                     comm,
+                                     &requests.back());
+          AssertThrowMPI(ierr);
+        }
+
+      // receive indices at the ranks we normally send from
+      for (const auto rank : send_ranks)
+        {
+          if (rank == my_rank)
+            continue;
+
+          MPI_Status status;
+          int        ierr = MPI_Probe(MPI_ANY_SOURCE, mpi_tag, comm, &status);
+          AssertThrowMPI(ierr);
+
+          int message_length;
+          ierr = MPI_Get_count(&status, MPI_CHAR, &message_length);
+          AssertThrowMPI(ierr);
+
+          std::vector<char> buffer(message_length);
+
+          ierr = MPI_Recv(buffer.data(),
+                          buffer.size(),
+                          MPI_CHAR,
+                          status.MPI_SOURCE,
+                          mpi_tag,
+                          comm,
+                          MPI_STATUS_IGNORE);
+          AssertThrowMPI(ierr);
+
+          indices_of_rank[status.MPI_SOURCE] =
+            Utilities::unpack<std::vector<unsigned int>>(buffer, false);
+        }
+
+      // make sure all messages have been sent
+      const int ierr =
+        MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
+      AssertThrowMPI(ierr);
+
+      return indices_of_rank;
+#endif
+    }
+
+
+
+    template <int structdim, int dim, int spacedim>
+    DistributedComputeIntersectionLocationsInternal<structdim, dim, spacedim>
+    distributed_compute_intersection_locations(
+      const GridTools::Cache<dim, spacedim> &          cache,
+      //TODO: check if std::vector instead of std::array can be used in ConsensusAlgorithms
```

Comment:
I guess we should support both `vector` and `array`? One first one for mixed meshes?

Reply:
@peterrum I initially used `std::vector<std::array>` to ensure the memory is continuous. If we use `std::vector<std::vector>` this is not ensured. As far as I understand, `MPI` this means one has to convert the data to a continuous (linear) `std::vector` (same holds for communicating the results). Is this valid? If so, do you know a place where it is done in `deal.II`?

Reply:
I guess the pack/unpack functions do the right thing, don't they?

Reply:
Can only use `std::vector` to begin with? In principle, `std::array` is easily supported, but then we need a huge amount of instantiations or can not put `distributed_compute_intersection_locations` in the source file.

Reply:
> Can only use std::vector to begin with?

:+1: